components/ChallengeMaterial.tsx

```diff
@@ -139,35 +152,131 @@ export const ChallengeTitleCard: React.FC<ChallengeTitleCardProps> = ({
   )
 }
 
-type ChallengeQuestionCardProps = {
+const ChallengeQuestionCardDisplay: React.FC<{
   currentChallenge: ChallengeSubmissionData
-}
-export const ChallengeQuestionCard: React.FC<ChallengeQuestionCardProps> = ({
-  currentChallenge
-}) => {
-  const diff = _.get(currentChallenge, 'submission.diff', '')
-  const comment = _.get(currentChallenge, 'submission.comment', '')
-  const updatedAt = _.get(currentChallenge, 'submission.updatedAt', Date.now())
+}> = ({ currentChallenge }) => {
+  const comment = currentChallenge.submission?.comment
+  const updatedAt = currentChallenge.submission?.updatedAt || ''
   const context = useContext(GlobalContext)
   const name = context.session?.user?.name
   const username = context.session?.user?.username
-  const reviewerUserName = _.get(
-    currentChallenge,
-    'submission.reviewer.username',
-    null
-  )
-  const comments =
-    currentChallenge.submission &&
-    currentChallenge.submission.comments?.filter(comment => !comment.line)
+  const userId = context.session?.user?.id
+
+  const [index, setIndex] = useState(-1)
+
+  const [submission, setSubmission] = useState(currentChallenge.submission)
+  const reviewerUserName = submission?.reviewer?.username || null
+  const comments = submission?.comments?.filter(comment => !comment.line)
 
   const [commentValue, setCommentValue] = React.useState('')
   const [addComment] = useAddCommentMutation()
-  const update = updateCache(
-    currentChallenge.submission?.id!,
-    commentValue,
-    name!,
-    username!
-  )
+
+  const { data, loading, error } = useGetPreviousSubmissionsQuery({
```

Comment:
Why not reduce this query to only get the past submissions ids and then fetch them on demand by its id? Querying them all here with the diff will increase the loading time when opening the review page (meaning the full screen loading spinner).

Reply:
I am not sure how it should work. GetPreviousSubmisisons query doesn't affect fullscreen loading spinner at all. 

Current fetching order is:
* getApp query (it should be in client's cache, and if this is not the case user is being redirected) 
* getSubmission query fetches all open submissions for this lesson (this is what actually hidden behind fullscreen loading spinner)
* Review page is rendered. getPreviousSubmissions query starts, select iteration in the top right corner is showing "Loading submissions...".
* getPreviousSubmissions returns and submission selection is fully rendered. So it doesn't affect loading spinner duration at all, it only increases time to final render but most users probably won't mind it (nobody notices a small component in the top right corner). 

Alternatives:
* Fetch everything in one query. First, I don't know how to do it (it's not possible in prisma and raw sql would be really big). Second, and more important this would actually increase loading spinner duration while reducing time to final render. This tradeoff would lead to bad UX in my opinion, why wait for full data when you can render it bit by bit and making page look more responsive?

* Use lazy loading. This again would be a worse UX. When you click on iteration you need to wait for query to return. Why not fetch it on first render while user is focused on reading initial diff?

Come to think about it, we need to get rid of fullscreen spinners someday. Nothing here prevents rendering of NavBar, Footer and Lesson components. Maybe even add a template loading submission just like youtube does it with gray rectangles.

Reply:
Yeah you're right, my mistake. We'll have to see how it will work out with big files like on JS6, the submission selector could take a longer time to load since it will have to fetch the entire submissions.

> Come to think about it, we need to get rid of fullscreen spinners someday. Nothing here prevents rendering of NavBar, Footer and Lesson components. Maybe even add a template loading submission just like youtube does it with gray rectangles.

Yes having a skeleton placeholder would be ideal.